upup/pkg/fi/cloudup/do/cloud.go

```diff
@@ -28,13 +42,526 @@ const TagNameEtcdClusterPrefix = "etcdCluster-"
 const TagNameRolePrefix = "k8s.io/role/"
 const TagKubernetesClusterNamePrefix = "KubernetesCluster"
 const TagKubernetesClusterMasterPrefix = "KubernetesCluster-Master"
+const TagKubernetesInstanceGroup = "kops-instancegroup"
 
-func SafeClusterName(clusterName string) string {
-	// DO does not support . in tags / names
-	safeClusterName := strings.Replace(clusterName, ".", "-", -1)
-	return safeClusterName
+type DOInstanceGroup struct {
+	ClusterName       string
+	InstanceGroupName string
+	GroupType         string   // will be either "master" or "worker"
+	Members           []string // will store the droplet names that matches.
 }
 
-func NewDOCloud(region string) (fi.Cloud, error) {
-	return digitalocean.NewCloud(region)
+// TokenSource implements oauth2.TokenSource
+type TokenSource struct {
+	AccessToken string
+}
+
+// DOCloud exposes all the interfaces required to operate on DigitalOcean resources
+type DOCloud interface {
+	fi.Cloud
+	//Client() *godo.Client
+	DropletsService() godo.DropletsService
+	DropletActionService() godo.DropletActionsService
+	VolumeService() godo.StorageService
+	VolumeActionService() godo.StorageActionsService
+	LoadBalancersService() godo.LoadBalancersService
+	DomainService() godo.DomainsService
+	ActionsService() godo.ActionsService
+	FindClusterStatus(cluster *kops.Cluster) (*kops.ClusterStatus, error)
+	GetAllLoadBalancers() ([]godo.LoadBalancer, error)
+	GetAllDropletsByTag(tag string) ([]godo.Droplet, error)
+	GetAllVolumesByRegion() ([]godo.Volume, error)
+}
+
+// static compile time check to validate DOCloud's fi.Cloud Interface.
+var _ fi.Cloud = &doCloudImplementation{}
+
+// doCloudImplementation holds the godo client object to interact with DO resources.
+type doCloudImplementation struct {
+	Client *godo.Client
+
+	dns dnsprovider.Interface
+
+	// region holds the DO region
```

Comment:
nit: should be a complete sentence, including punctuation:

```suggestion
	// region holds the DO region.
```

Reply:
Sure, updated.